src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -56,43 +57,90 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
             {
-                if (PathUtility.GetStringComparerBasedOnOS().Equals(p2p.RestoreMetadata.ProjectUniqueName, uniqueName))
+                var target = nuGetLockFile.Targets.FirstOrDefault(
+                    t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+
+                if (target == null)
                 {
-                    continue;
+                    // a new target found in the dgSpec so invalidate existing lock file.
+                    return false;
                 }
 
-                var projectName = Path.GetFileNameWithoutExtension(p2p.RestoreMetadata.ProjectPath);
+                var queue = new Queue<string>();
+                var visitedP2PReference = new HashSet<string>();
 
-                foreach (var framework in p2p.TargetFrameworks)
+                foreach (var projectReference in framework.ProjectReferences)
                 {
-                    var target = nuGetLockFile.Targets.FirstOrDefault(
-                    t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
-
-                    if (target != null)
+                    if (visitedP2PReference.Add(projectReference.ProjectUniqueName))
                     {
-                        var projectDependency = target.Dependencies.FirstOrDefault(
-                            dep => dep.Type == PackageDependencyType.Project &&
-                            PathUtility.GetStringComparerBasedOnOS().Equals(dep.Id, projectName));
+                        queue.Enqueue(projectReference.ProjectUniqueName);
 
-                        if (HasP2PDependencyChanged(framework.Dependencies, projectDependency))
+                        while (queue.Count > 0)
                         {
-                            // lock file is out of sync
-                            return false;
+                            var p2pUniqueName = queue.Dequeue();
+                            var p2pProjectName = Path.GetFileNameWithoutExtension(p2pUniqueName);
+
+                            var projectDependency = target.Dependencies.FirstOrDefault(
+                                dep => dep.Type == PackageDependencyType.Project &&
+                                PathUtility.GetStringComparerBasedOnOS().Equals(dep.Id, p2pProjectName));
+
+                            if (projectDependency == null)
+                            {
+                                // project dependency doesn't exist in lock file.
+                                return false;
+                            }
+
+                            var p2pSpec = dgSpec.GetProjectSpec(p2pUniqueName);
+
+                            // ignore if this p2p packageSpec not found in DGSpec, It'll fail restore with different error at later stage
+                            if (p2pSpec != null)
+                            {
+                                var p2pSpecTarget = p2pSpec.TargetFrameworks.FirstOrDefault(
```

Comment:
I don't know if you want first or default here, I think you want the best framework which is what normal restore does.

Reply:
this is a ordered list of frameworks where we go with the first compatible framework. This is consistent with restore since it also uses the first compatible framework of P2P or package dependencies.

Reply:
Let's discuss this offline.

Reply:
I did some basic tests and I did not find that to be the case. 

A (net46,net45) -> B(net45, net46). 

The graph is A (net46) -> B(net46) and A(net45) -> B(net45)

Reply:
On my machine if I replaace that line with:
```
var p2pSpecTarget = NuGetFrameworkUtility.GetNearest(p2pSpec.TargetFrameworks, framework.FrameworkName, e => e.FrameworkName);
```
it passes

Reply:
I still don't understand why it's passing on my box but tried on another machine, and it threw the same error as yours. This is anyways a safe change and without wasting more time, I'll change it to get nearest framework.

Reply:
Updated the PR with this change and also added a test for the same scenario.